Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_wfsi.cpp

```diff
@@ -0,0 +1,290 @@
+// Copyright 2016 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "Core/IPC_HLE/WII_IPC_HLE_Device_wfsi.h"
+
+#include <mbedtls/aes.h>
+#include <stack>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "Common/CommonTypes.h"
+#include "Common/FileUtil.h"
+#include "Common/Logging/Log.h"
+#include "Core/HW/Memmap.h"
+#include "Core/IPC_HLE/WII_IPC_HLE_Device_usb_wfssrv.h"
+#include "DiscIO/NANDContentLoader.h"
+
+void ARCUnpacker::Reset()
+{
+  m_whole_file.clear();
+}
+
+void ARCUnpacker::AddBytes(const std::vector<u8>& bytes)
+{
+  m_whole_file.insert(m_whole_file.end(), bytes.begin(), bytes.end());
+}
+
+void ARCUnpacker::Extract(const WriteCallback& callback)
+{
+  u32 fourcc = Common::swap32(m_whole_file.data());
+  if (fourcc != 0x55AA382D)
+  {
+    ERROR_LOG(WII_IPC_HLE, "ARCUnpacker: invalid fourcc (%08x)", fourcc);
+    return;
+  }
+
+  // Read the root node to get the number of nodes.
+  u8* nodes_directory = m_whole_file.data() + 0x20;
+  u32 nodes_count = Common::swap32(nodes_directory + 8);
+  constexpr u32 NODE_SIZE = 0xC;
+  char* string_table = reinterpret_cast<char*>(nodes_directory + nodes_count * NODE_SIZE);
+
+  std::stack<std::pair<u32, std::string>> directory_stack;
+  directory_stack.emplace(std::make_pair(nodes_count, ""));
+  for (u32 i = 1; i < nodes_count; ++i)
+  {
+    while (i >= directory_stack.top().first)
+    {
+      directory_stack.pop();
+    }
+    const std::string& current_directory = directory_stack.top().second;
+    u8* node = nodes_directory + i * NODE_SIZE;
+    u32 name_offset = (node[1] << 16) | Common::swap16(node + 2);
+    u32 data_offset = Common::swap32(node + 4);
+    u32 size = Common::swap32(node + 8);
+    std::string basename = string_table + name_offset;
+    std::string fullname =
+        current_directory.empty() ? basename : current_directory + "/" + basename;
+
+    u8 flags = *node;
+    if (flags == 1)
+    {
+      directory_stack.emplace(std::make_pair(size, fullname));
+    }
+    else
+    {
+      std::vector<u8> contents(m_whole_file.data() + data_offset,
+                               m_whole_file.data() + data_offset + size);
+      callback(fullname, contents);
+    }
+  }
+}
+
+CWII_IPC_HLE_Device_wfsi::CWII_IPC_HLE_Device_wfsi(u32 device_id, const std::string& device_name)
+    : IWII_IPC_HLE_Device(device_id, device_name)
+{
+  memset(m_aes_key, 0, sizeof(m_aes_key));
+  memset(m_aes_iv, 0, sizeof(m_aes_iv));
+}
+
+IPCCommandResult CWII_IPC_HLE_Device_wfsi::Open(u32 command_address, u32 mode)
+{
+  INFO_LOG(WII_IPC_HLE, "/dev/wfsi: Open");
+  Memory::Write_U32(GetDeviceID(), command_address + 4);
```

Comment:
This isn't actually needed, see https://github.com/dolphin-emu/dolphin/commit/27e63632146a2252b56fd0bdc853dfe76bffecd2